src/main/java/com/eighthlight/fabrial/App.java

```diff
@@ -0,0 +1,47 @@
+package com.eighthlight.fabrial;
+
+import com.eighthlight.fabrial.server.ServerConfig;
+import com.eighthlight.fabrial.server.TcpServer;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class App {
+  static final Logger logger = Logger.getLogger(App.class.getName());
+
+  static TcpServer server;
+
+  public static void start() {
+    server = new TcpServer(new ServerConfig());
+    try {
+      server.start();
+    } catch (IOException e) {
+      logger.log(Level.SEVERE, "Failed to start server", e);
+      System.exit(1);
+    }
+    Runtime.getRuntime().addShutdownHook(new Thread(App::shutdown));
+  }
+
+  public static void run() {
+    while (!server.isClosed()) {
```

Comment:
I suspect this will be pretty hot on your CPU. Might be interesting to check it out & see. 

If it is burning CPU in this tight loop, a few alternative ideas:
- add a `Thread.sleep(...)` to let it relax for awhile
- make `server.start()` blocking, and remove the `run()` method entirely - and when you need it not to block, e.g. in the acceptance tests, call it in a separate thread/executor
- make sure the server's thread of execution is a "non-daemon" thread, which will prevent JVM shutdown until that thread exits (it may already be that way! it's the default I think) - your `TcpServer.acceptConnections` already has a loop that will complete (and thus exit that thread) when the `serverSocket` closes.

Reply:
I'll take your word for it 😉 Either way it's not very elegant. Here's what I'm thinking

- I like that `start` is async, since it's easier to call `close`, but that might not make sense in the actual `App` use case
- Perhaps I can have a method `awaitClose` which waits until the `acceptThread` exits
- I'll double check that `acceptThread` (and the connection handling threads) aren't daemonized

Reply:
Oh sorry, I hadn't checked locally - but I just did, and confirmed one CPU was pegged.

What I was trying to say in my last item is that I think you ought to be able to remove this `run()` method entirely - since your shutdown hook is taking care of your logging on exit, and a non-daemon running `TcpServer.acceptConnections` would prevent JVM shutdown until it completes. So the `awaitClose` wouldn't be necessary either. 

I just tried this locally as well (deleting `run()` and the "Goodbye!" logging), and the shutdown process appears to be identical to before - the server does not immediately exit, but serves requests normally - until I kill it with <ctrl>-c, at which point normal shutdown happens.

Reply:
Oh, interesting. I assumed _something_ had to be happening in `main` to keep it alive. Good to know the "parent" process lives as long non-daemon threads are running.